### PR TITLE
[Documentation] Customization Guide: Factories

### PR DESCRIPTION
**Pull request by TheMadeleine**
| Q               | A
| --------------- | ---
| Doc fix?        | no |
| New docs?    | yes |
| Related tickets | partially #4226, mentioned in #7004 |
| License         | MIT |

### CODE REVIEW

File: docs/customization/factory.rst [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+Let's assume that you would want to have a possibility to create disabled products.
+
+1. Create your own factory class in the ``AppBundle\Factory`` namespace.
+Remember that it has to extend a proper base class. How can you check that?
+
+For the ``ProductFactory`` run:
+
+.. code-block:: bash
+
+    $ php bin/console debug:container sylius.factory.product
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: Shouldn't we use our custom debugger? `sylius:debug:resource sylius.product`

File: docs/customization/factory.rst [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+As a result you will get the ``Sylius\Component\Product\Factory\ProductFactory`` - this is the class that you need to be extending.
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: We will receive a table actually. Something like: `Check class option. It is value is a class that we are interested in`

File: docs/customization/factory.rst [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+    namespace AppBundle\Factory;
+
+    use Sylius\Component\Core\Model\AddressInterface;
+    use Sylius\Component\Core\Model\ProductInterface;
+    use Sylius\Component\Product\Factory\ProductFactory as BaseProductFactory;
+    use Sylius\Component\Core\Model\CustomerInterface;
+    use Sylius\Component\Resource\Factory\FactoryInterface;
+
+    class ProductFactory extends BaseProductFactory
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: We shouldn't extend, but decorate. Wrong pattern. Also, it should implements `ProductFactoryInterface`

File: docs/customization/factory.rst [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        /**
+         * @var FactoryInterface
+         */
+        private $factory;
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: decoratedFacotry

File: docs/customization/factory.rst [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+        /**
+         * @var FactoryInterface
+         */
+        private $variantFactory;
+
+        /**
+         * @param FactoryInterface $factory
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: Only one parameter is needed with `ProductFactoryInterface` typehint

File: docs/customization/factory.rst [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            $product = $this->decoratedFactory->createNew();
+
+            $product->setEnabled(false);
+
+            return $product;
+        }
+    }
+
+2. In order to decorate the base ProductFactory with your implementation you need to configure it
+as a decorating service in the ``AppBundle\Resources\config\services.yml``.
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: `AppBundle\Resources\config\services.yml` -> `app/config/services.yml`

File: docs/customization/factory.rst [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+.. code-block:: yaml
+
+    services:
+        app.factory.product:
+            class: AppBundle\Factory\ProductFactory
+            decorates: sylius.factory.product
+            arguments: ['@app.factory.product.inner']
+            public: false
```
Comment (lchrusciel) [inline docs/customization/factory.rst]: 👍

File: docs/customization/factory.rst [hunk 8]
```diff
@@ -0,0 +1,131 @@
+Customizing Factories
+=====================
+
+.. warning::
+
+    In **Sylius** we are already decorating factories from components in Core.
+    Often you will be needing to add your very own methods to them. You need to check before which factory is your resource using.
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: Scratch the first sentence - it's irrelevant in the warning scope.
Furthermore, the why, and how are explained right below.
`before which` => `beforehand, which`

File: docs/customization/factory.rst [hunk 9]
```diff
@@ -0,0 +1,131 @@
+Customizing Factories
+=====================
+
+.. warning::
+
+    In **Sylius** we are already decorating factories from components in Core.
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: How 'bout `Some factories may already be decorated in the **Sylius** Core.`

File: docs/customization/factory.rst [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Differently configured versions of resources may be needed in various scenarios in your application.
+You may need for instance to:
+
+    * create Product with Supplier(which is your own custom entity)
+    * create a disabled Product(for further modifications)
+    * create a ProductReview with predefined description
+
+and many, many more usecases.
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: `usecases` is redundant.

File: docs/customization/factory.rst [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+How to customize a Factory?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's assume that you would want to have a possibility to create disabled products.
+
+1. Create your own factory class in the ``AppBundle\Factory`` namespace.
+Remember that it has to extend a proper base class. How can you check that?
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: ⚠️  `extend` ⚠️

File: docs/customization/factory.rst [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+For the ``ProductFactory`` run:
+
+.. code-block:: bash
+
+    $ php bin/console sylius:debug:resource sylius.product
+
+As a result you will get a table od Product related classes. Check the ``classes.factory`` row
+where you will find the ``Sylius\Component\Product\Factory\ProductFactory`` - this is the class that you need to be decorating.
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: `(...) you need to decorate.` ?

File: docs/customization/factory.rst [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+.. code-block:: yaml
+
+    services:
+        app.factory.product:
+            class: AppBundle\Factory\ProductFactory
+            decorates: sylius.factory.product
+            arguments: ['@app.factory.product.inner']
+            public: false
+
+3. After the ``sylius.factory.product`` has been decorated it has been extended by the new ``createDisabled()`` method.
```
Comment (NeverResponse) [inline docs/customization/factory.rst]: Points **1.** and **2.** tell people what needs to be done, this one however is purely informational. Would be  more natural, if lines 104 and 106 were switched, with a slight rephrasing.

**Comment by pjedrzejewski** [comment]
Thanks Magda, nice work! And thank you all for the review! ;)

**Pull request merged**